torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -1032,59 +1939,76 @@ void SegmentCandidateFinder::findSegments() {
+  if (options_.run_combine_reductions && CombineReductions::shouldRun(this)) {
+    CombineReductions::run(this);
+  }
 
-  // Initial merge iteration
-  while (merged_nodes) {
-    // Reset stateful traversal details in SegmentedGroups
-    resetTraversal();
+  // All merges will be vertical beyond this point for now, so
+  //  we can remove the input auxiliary groups. Should make the vertical
+  //  merges avoid auxiliary group once we start general horizontal merges
+  std::unordered_set<SegmentedGroup*> input_groups;
+  for (auto input : completeFusion().inputs()) {
+    input_groups.insert(input2group.at(input));
+  }
+  eraseGroups(input_groups);
 
-    resetLevels();
+  if (options_.run_herrmann_merge) {
+    bool merged_nodes = true;
+    // Initial merge iteration
+    while (merged_nodes) {
+      // Reset stateful traversal details in SegmentedGroups
+      resetTraversal();
 
-    for (auto& group : groups()) {
-      if (group->merged_) {
-        continue;
-      }
-      auto candidates = group->getMergeCandidates();
-      if (candidates.empty()) {
-        continue;
-      }
+      resetLevels();
 
-      auto candidate_it = candidates.begin();
-      while (candidate_it != candidates.end() &&
-             !codeGenSupportedMerge(candidate_it->edge)) {
-        candidate_it++;
-      }
-      if (candidate_it == candidates.end()) {
-        continue;
-      }
+      for (auto& group : groups()) {
+        if (group->merged_) {
+          continue;
+        }
+        auto candidates = group->getMergeCandidates();
+        if (candidates.empty()) {
+          continue;
+        }
+
+        auto candidate_it = candidates.begin();
+        while (candidate_it != candidates.end() &&
+               !codeGenSupportedMerge(candidate_it->edge)) {
+          candidate_it++;
+        }
+        if (candidate_it == candidates.end()) {
+          continue;
+        }
 
-      to_merge_.emplace_back(group);
-      to_merge_.emplace_back(candidate_it->group);
+        to_merge_.emplace_back(group);
+        to_merge_.emplace_back(candidate_it->group);
 
-      group->merged_ = true;
-      group->merge_with_ = candidate_it->group;
-      group->merge_through_ = candidate_it->edge;
+        group->merged_ = true;
+        group->merge_with_ = candidate_it->group;
+        group->merge_through_ = candidate_it->edge;
 
-      candidate_it->group->merged_ = true;
-      candidate_it->group->merge_with_ = group;
-      candidate_it->group->merge_through_ = candidate_it->edge;
-    }
+        candidate_it->group->merged_ = true;
+        candidate_it->group->merge_with_ = group;
+        candidate_it->group->merge_through_ = candidate_it->edge;
+      }
 
-    if (to_merge_.empty()) {
-      merged_nodes = false;
-    }
+      if (to_merge_.empty()) {
+        merged_nodes = false;
+      }
 
-    mergeNodes();
+      mergeNodes();
+    }
   }
 
-  finalMerge();
+  if (options_.run_final_merge) {
+    finalMerge();
+  }
```

Comment:
We may want to consider ping-ponging between final merge and the Herrmann iterations, as the "final merge" might actually create opportunities for the Herrmann iterations to progress. If we did that, should rename final merge to brute force search. Doesn't need to be in this PR, just something to consider or mark as todo to consider.